File: src/main/java/com/mpe85/grampa/rule/impl/TrieRule.java
package com.mpe85.grampa.rule.impl;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.Objects;
import java.util.Set;

import com.google.common.base.MoreObjects.ToStringHelper;
import com.google.common.collect.Sets;
import com.ibm.icu.lang.UCharacter;
import com.ibm.icu.util.BytesTrie.Result;
import com.ibm.icu.util.CharsTrie;
import com.ibm.icu.util.CharsTrieBuilder;
import com.ibm.icu.util.StringTrieBuilder.Option;
import com.mpe85.grampa.rule.RuleContext;

import one.util.streamex.StreamEx;

public class TrieRule<T> extends AbstractRule<T> {
	
	public TrieRule(final String... strings) {
		this(false, strings);
	}
	
	public TrieRule(final boolean ignoreCase, final String... strings) {
		this(Sets.newHashSet(strings), ignoreCase);
	}
	
	public TrieRule(final Set<String> strings) {
		this(strings, false);
	}
	
	public TrieRule(
			final Set<String> strings,
			final boolean ignoreCase) {
		final CharsTrieBuilder builder = new CharsTrieBuilder();
		strings.forEach(s -> {
			checkNotNull(s, "A 'string' must not be null.");
			builder.add(ignoreCase ? UCharacter.toLowerCase(s) : s, 0);
		});
		trie = builder.build(Option.FAST);
		this.ignoreCase = ignoreCase;
	}
	
	public Set<String> getStrings() {
		return StreamEx.of(trie.iterator()).map(e -> e.chars).map(String.class::cast).toSet();
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		int longestMatch = 0;
		final int[] codePoints = context.getRestOfInput().codePoints().toArray();
		for (int i = 0; i < codePoints.length; i++) {
			final Result result = trie.next(ignoreCase ? UCharacter.toLowerCase(codePoints[i]) : codePoints[i]);
			if (result == Result.FINAL_VALUE || result == Result.INTERMEDIATE_VALUE) {
				longestMatch = i + 1;
			}
			if (result == Result.FINAL_VALUE || result == Result.NO_MATCH) {
				break;
			}
		}
		trie.reset();
		return longestMatch > 0 && context.advanceIndex(longestMatch);
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), getStrings(), ignoreCase);
	}
	
	@Override
	public boolean equals(final Object obj) {
		if (obj != null && getClass() == obj.getClass()) {
			final TrieRule<?> other = (TrieRule<?>) obj;
			return super.equals(other)
					&& Objects.equals(getStrings(), other.getStrings())
					&& Objects.equals(ignoreCase, other.ignoreCase);
		}
		return false;
	}
	
	@Override
	protected ToStringHelper toStringHelper() {
		return super.toStringHelper()
				.add("strings", getStrings())
				.add("ignoreCase", ignoreCase);
	}
	
	
	private final CharsTrie trie;
	private final boolean ignoreCase;
}

File: src/main/kotlin/com/mpe85/grampa/rule/impl/TrieRule.kt
package com.mpe85.grampa.rule.impl

import com.ibm.icu.lang.UCharacter
import com.ibm.icu.lang.UCharacter.charCount
import com.ibm.icu.lang.UCharacter.toString
import com.ibm.icu.util.BytesTrie.Result.FINAL_VALUE
import com.ibm.icu.util.BytesTrie.Result.INTERMEDIATE_VALUE
import com.ibm.icu.util.BytesTrie.Result.NO_MATCH
import com.ibm.icu.util.CharsTrieBuilder
import com.ibm.icu.util.StringTrieBuilder.Option.FAST
import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.util.checkEquality
import com.mpe85.grampa.util.stringify
import java.util.Objects.hash
import kotlin.streams.asSequence
import kotlin.streams.toList

/**
 * A case-sensitive trie (prefix tree) rule implementation that matches the input against a dictionary.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[strings] A collection of strings from which the trie is built up
 */
public open class TrieRule<T> constructor(private val strings: Collection<String>) : AbstractRule<T>() {

    private val trie = CharsTrieBuilder().run {
        strings.map(::map).distinct().forEach { add(it, 0) }
        build(FAST)
    }

    protected open fun map(string: String): String = string

    /**
     * Construct a case-sensitive trie.
     *
     * @param[strings] A variable number of strings
     */
    public constructor(vararg strings: String) : this(strings.toList())

    override fun match(context: ParserContext<T>): Boolean = try {
        var longestMatch = 0
        for ((idx, codePoint) in context.restOfInput.codePoints().asSequence().withIndex()) {
            val foldedCodePoints = map(toString(codePoint)).codePoints().toList()
            foldedCodePoints.dropLast(1).forEach { trie.nextForCodePoint(it) }
            val result = trie.nextForCodePoint(foldedCodePoints.last())
            if (result in setOf(FINAL_VALUE, INTERMEDIATE_VALUE)) {
                longestMatch = idx + 1
            }
            if (result in setOf(FINAL_VALUE, NO_MATCH)) {
                break
            }
        }
        val charCount = context.restOfInput.codePoints().asSequence().take(longestMatch).sumBy { charCount(it) }
        longestMatch > 0 && context.advanceIndex(charCount)
    } finally {
        trie.reset()
    }

    override fun hashCode(): Int = hash(super.hashCode(), strings)
    override fun equals(other: Any?): Boolean = checkEquality(other, { super.equals(other) }, { it.strings })
    override fun toString(): String = stringify("strings" to strings)
}

/**
 * A case-insensitive trie (prefix tree) rule implementation that matches the input against a dictionary.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[strings] A collection of strings from which the trie is built up
 */
public class IgnoreCaseTrieRule<T>(strings: Collection<String>) : TrieRule<T>(strings) {

    /**
     * Construct a case-insensitive trie.
     *
     * @param[strings] A variable number of strings
     */
    public constructor(vararg strings: String) : this(strings.toList())

    override fun map(string: String): String = UCharacter.foldCase(string, true)

}

